### PR TITLE
Introducing Collectors - final step :partying_face:  - processing collected data + test case

### PR DESCRIPTION
**Pull request by TomasVotruba**
Test with following command:

```bash
bin/rector process src/Application -n --clear-cache
```

Inspiration from PHPStan:

* core - https://github.com/phpstan/phpstan-src/commit/bccd1cd58e0d117ac8755ab228e338d966cac25a
* tests - https://github.com/phpstan/phpstan-src/commit/b10fb30b2b1f48a9615f6a464acf475adb992d4a

### CODE REVIEW

**Comment by TomasVotruba** [comment]
This PR adds support for collectors, same as in PHPStan.
It's wip and possibly will narrow down to `AbstractRector` and single contract, similar to PHPStan.

Any ideas to improve design are welcomed :pray: 
This is ready for the feedback here @staab @samsonasik  :)

File: packages/Parallel/ValueObject/Bridge.php [hunk 1]
```diff
@@ -37,5 +37,10 @@ final class Bridge
     /**
      * @var string
      */
-    public const COLLECTED_DATA = 'collected_data';
+    public const COLLECTED_DATAS = 'collected_datas';
```
Comment (samsonasik) [inline packages/Parallel/ValueObject/Bridge.php]: afaik, data is already a collection of datum, so i prefer data instead of datas

File: packages/Config/RectorConfig.php [hunk 2]
```diff
@@ -192,15 +195,30 @@ public function ruleWithConfiguration(string $rectorClass, array $configuration)
...
-        Assert::isAOf($rectorClass, RectorInterface::class);
+        Assert::isAnyOf($rectorClass, [RectorInterface::class, CollectorRectorInterface::class]);
 
         $this->singleton($rectorClass);
-        $this->tag($rectorClass, RectorInterface::class);
+        if (is_a($rectorClass, RectorInterface::class, true)) {
+            $this->tag($rectorClass, RectorInterface::class);
+        } elseif (is_a($rectorClass, CollectorRectorInterface::class, true)) {
+            $this->tag($rectorClass, CollectorRectorInterface::class);
+        } else {
```
Comment (staabm) [inline packages/Config/RectorConfig.php]: I would prefer `instanceof` over `is_a`

File: packages/Parallel/Application/ParallelFileProcessor.php [hunk 3]
```diff
@@ -107,9 +117,11 @@ public function process(
                 }
 
                 $jobsChunk = array_pop($jobs);
+
                 $parallelProcess->request([
                     ReactCommand::ACTION => Action::MAIN,
                     Content::FILES => $jobsChunk,
+                    Bridge::PREVIOUSLY_COLLECTED_DATAS => $configuration->getCollectedDatas(),
```
Comment (staabm) [inline packages/Parallel/Application/ParallelFileProcessor.php]: minor: `DATA` is already plural.

File: rules/Php80/DocBlock/PropertyPromotionDocBlockMerger.php [hunk 4]
```diff
@@ -65,8 +65,6 @@ public function decorateParamWithPropertyPhpDocInfo(
         string $paramName
     ): void {
         $propertyPhpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($property);
-        $propertyPhpDocInfo->markAsChanged();
```
Comment (staabm) [inline rules/Php80/DocBlock/PropertyPromotionDocBlockMerger.php]: unrelated change?

File: src/ValueObject/Error/SystemError.php [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 namespace Rector\Core\ValueObject\Error;
 
-use Rector\Parallel\ValueObject\Name;
+use Rector\Parallel\ValueObject\BridgeItem;
```
Comment (staabm) [inline src/ValueObject/Error/SystemError.php]: it might make sense to separate the `BridgeItem` rename into a separate PR which can land before collectors

**Review by staabm — COMMENTED** [review]
a first pass

File: rules/Php80/DocBlock/PropertyPromotionDocBlockMerger.php (hunk 4 above)
Comment (samsonasik) [reply rules/Php80/DocBlock/PropertyPromotionDocBlockMerger.php]: Yup

**Comment by staabm** [comment]
if I get it right, to utilize a PHPStan-Collector we implement a class which implements `PHPStan\Collectors\Collector` and this collector will be auto-discovered by rector? no need to configure said collector-class - like it is in PHPStan in .neon files?

----

I miss the part where a rector rule is subscribing itself to collected data, like we do in PHPStan with

```
public function getNodeType(): string
{
	return CollectedDataNode::class;
}
```


said differently: from a point of view as someone who knows PHPStan Collectors, it feels weird that the same classes are used, but the mechanics are different.

if I see it correctly a rector rule is subscribing to `CollectedDataNode` by implemeting the `CollectorRectorInterface` interface?

---

the above are mostly DX concerns. In generell the combination of rector and collectors is awesome 🥳

File: src/Application/Collector/CollectorNodeVisitor.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        return null;
+    }
+
+    public function enterNode(Node $node)
+    {
+        $collectors = $this->collectorRegistry->getCollectors($node::class);
+
+        foreach ($collectors as $collector) {
+            /** @var Scope $scope */
+            $scope = $node->getAttribute(AttributeKey::SCOPE);
```
Comment (staabm) [inline src/Application/Collector/CollectorNodeVisitor.php]: move before loop

File: src/Application/FileProcessor.php [hunk 7]
```diff
@@ -50,6 +52,11 @@ public function __construct(
 
     public function processFile(File $file, Configuration $configuration): FileProcessResult
     {
+        if ($configuration->isSecondRun()) {
+            // 2nd run
+            $this->rectorNodeTraverser->prepareCollectorRectorsRun($configuration);
```
Comment (staabm) [inline src/Application/FileProcessor.php:57]: does it mean, on the "2nd run" we only run CollectorRectors, and on the 1st run, we run regular-rectors and data-collectors at the same time?

if so, I am wondering whether regular-rectors which modify the AST can lead to invalid/no longer up 2 date data beeing collected by the collectors

File: packages/Parallel/ValueObject/Bridge.php (hunk 1 above)
Comment (TomasVotruba) [reply packages/Parallel/ValueObject/Bridge.php]: I see. I had to change it to iterate with single naming.

How would you do singular from that?
Maybe different naming?

File: packages/Config/RectorConfig.php (hunk 2 above)
Comment (TomasVotruba) [reply packages/Config/RectorConfig.php]: These are both strings, so not sure if that works.

File: packages/Parallel/ValueObject/Bridge.php (hunk 1 above)
Comment (samsonasik) [reply packages/Parallel/ValueObject/Bridge.php]: collected_data should be ok:

```suggestion
    public const COLLECTED_DATA = 'collected_data';
```

then, on loop, use `$datum` as singular:

```php
foreach ($data[Bridge::COLLECTED_DATA as $datum)
```

File: packages/Config/RectorConfig.php (hunk 2 above)
Comment (staabm) [reply packages/Config/RectorConfig.php]: ahh right. I fall into this trap over and over ;)

File: packages/Parallel/ValueObject/Bridge.php (hunk 1 above)
Comment (TomasVotruba) [reply packages/Parallel/ValueObject/Bridge.php]: "datum" sounds like a czech version and I haven't heard it before. I'd prefer going with something well known :)

File: packages/Config/RectorConfig.php (hunk 2 above)
Comment (TomasVotruba) [reply packages/Config/RectorConfig.php]: I think we have a Rector rule for your case already, so no need to check this :laughing: 

https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#inlineisainstanceofrector

I'm rather interested in architecture and idea feedback :pray:

File: src/ValueObject/Error/SystemError.php (hunk 5 above)
Comment (TomasVotruba) [reply src/ValueObject/Error/SystemError.php]: Indeed, I'm on it

File: packages/Parallel/ValueObject/Bridge.php (hunk 1 above)
Comment (staabm) [reply packages/Parallel/ValueObject/Bridge.php]: `foreach ($data[Bridge::COLLECTED_DATA as $dataItem)`

or

`foreach ($data[Bridge::COLLECTED_DATA as $item)`

File: src/Application/Collector/CollectorNodeVisitor.php (hunk 6 above)
Comment (TomasVotruba) [reply src/Application/Collector/CollectorNodeVisitor.php]: Thanks :+1:

File: src/Application/FileProcessor.php (hunk 7 above)
Comment (staabm) [reply src/Application/FileProcessor.php:57]: since you put a :+1: on it, let me explain further:

I think letting collectors run, while regular rector are still beeing applied and move things arround is wasteful.
wouldn't it make sense to apply all regular rectors first. after the changes stabilized let the collectors run and after that run the collectorrectors (and after that start over)? 

or is this something PHPStan pre-defines and we can't influence?

---

in native PHPStan "its pretty easy" to just run all collectors and afterwards run all rules, because the code will not change in between. but in our case, every rector-rule can influence the ast and therefore we might need to start all over a lot and therefore need to make sure the every changing ast is properly reflected in the collectors and all rector rules always work on the latest AST (and everything is still fast and efficient)

**Comment by staabm** [comment]
one more question: do you think, utilizing the collector infrastructure we might be able to get rid of the scope-attribute on nodes, as rectors which require this additional scope context atm can use collectors instead so we don't need AST traversal in regular rectors anymore?

architecture wise having the scope attribute on each node is the biggest problem in the rector-src codebase IMO

File: rules/TypeDeclaration/Collector/ParentClassCollector.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Rector\TypeDeclaration\Collector;
+
+use PhpParser\Node;
+use PhpParser\Node\Name;
+use PhpParser\Node\Stmt\Class_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Collectors\Collector;
+
+final class ParentClassCollector implements Collector
```
Comment (staabm) [inline rules/TypeDeclaration/Collector/ParentClassCollector.php:16]: should have a `@implements<Class_, class-string[]>` generic, so PHPStan knows what data will be collected and on the cosuming side is aware which data can be fetched from the collector

Comment (TomasVotruba) [reply rules/TypeDeclaration/Collector/ParentClassCollector.php:16]: Done :+1:

Comment (TomasVotruba) [reply rules/TypeDeclaration/Collector/ParentClassCollector.php:16]: Hm, when I add it, it yield some contravariant bug I've never understood :D could you add it so it works well?

Comment (staabm) [reply rules/TypeDeclaration/Collector/ParentClassCollector.php:16]: it has to be

```diff
$ git diff
diff --git a/rules/TypeDeclaration/Collector/ParentClassCollector.php b/rules/TypeDeclaration/Collector/ParentClassCollector.php
index ad73832c5d..48125ca462 100644
--- a/rules/TypeDeclaration/Collector/ParentClassCollector.php
+++ b/rules/TypeDeclaration/Collector/ParentClassCollector.php
@@ -10,6 +10,9 @@ use PhpParser\Node\Stmt\Class_;
 use PHPStan\Analyser\Scope;
 use PHPStan\Collectors\Collector;

+/**
+ * @implements Collector<Class_, string[]>
+ */
 final class ParentClassCollector implements Collector
 {
     public function getNodeType(): string
@@ -19,7 +22,6 @@ final class ParentClassCollector implements Collector

     /**
      * @param Class_ $node
-     * @return mixed[]|null
      */
     public function processNode(Node $node, Scope $scope): ?array
     {

```

Comment (TomasVotruba) [reply rules/TypeDeclaration/Collector/ParentClassCollector.php:16]: Ah, got it :) thanks :+1:

File: src/Application/FileProcessor.php (hunk 7 above)
Comment (TomasVotruba) [reply src/Application/FileProcessor.php:57]: > if so, I am wondering whether regular-rectors which modify the AST can lead to invalid/no longer up 2 date data beeing collected by the collectors

The collectors only collect metadata, not nodes. E.g. in this PR looks for classes, that are extended. These classes cannot be final then.

On 2nd run only the `CollectorRectorInterface` will iterate over nodes and only finalize classes, that are not extended.

Comment (TomasVotruba) [reply src/Application/FileProcessor.php:57]: > wouldn't it make sense to apply all regular rectors first. after the changes stabilized let the collectors run and after that run the collectorrectors (and after that start over)?

I think that's how it's done now:

* 1st run only normal Rectors + collect metadata
* 2nd run use collected metadata and run only Rector rules that will modify nodes based on those metadata

**Comment by TomasVotruba** [comment]
> one more question: do you think, utilizing the collector infrastructure we might be able to get rid of the scope-attribute on nodes, as rectors which require this additional scope context atm can use collectors instead so we don't need AST traversal in regular rectors anymore?

At the moment running Rector parsing twice is twice as costly in time and possibly memory, so I'd reserve these rules only to cases where they're needed. The scope-attribute should be still used to cases it deals with best.

**Comment by staabm** [comment]
> The scope-attribute should be still used to cases it deals with best.

my gut feeling is, that the scope attribute is super expensive memory wise. I need to do a small experiement to prove that :)

File: src/Application/FileProcessor.php (hunk 7 above)
Comment (staabm) [reply src/Application/FileProcessor.php:57]: > The collectors only collect metadata, not nodes. E.g. in this PR looks for classes, that are extended. These classes cannot be final then.
 

I don't think so. the collectors can return whatever they like .. it might collect AST nodes

**Comment by TomasVotruba** [comment]
@staabm Sure, that's another topic outside this PR scope :)

**Comment by staabm** [comment]
This looks good enough to play with it

**Comment by TomasVotruba** [comment]
Let's ship it to test out in the wild :lion:

**Comment by staabm** [comment]
great job! 👏

**Comment by TomasVotruba** [comment]
Thanks :muscle:  It was a challange to figure out the order and workflow. Let's see how it goes in the wild :)

File: packages/Config/RectorConfig.php [hunk 9]
```diff
@@ -59,6 +60,14 @@ public function disableParallel(): void
         SimpleParameterProvider::setParameter(Option::PARALLEL, false);
     }
 
+    /**
+     * @experimental since Rector 0.17.x
```
Comment (Wirone) [inline packages/Config/RectorConfig.php:64]: Should be `since Rector 0.18.4` or at least `0.18.x` 😉

File: src/Application/Collector/CollectorNodeVisitor.php [hunk 10]
```diff
@@ -66,7 +66,7 @@ public function enterNode(Node $node)
     /**
      * @return CollectedData[]
      */
-    public function getCollectedDatas(): array
+    public function getCollectedData(): array
     {
         return $this->collectedDatas;
```
Comment (Wirone) [inline src/Application/Collector/CollectorNodeVisitor.php:71]: `CollectorNodeVisitor::$collectedDatas` should be changed to singular form too.

File: src/Console/Command/ProcessCommand.php [hunk 11]
```diff
@@ -83,8 +83,27 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+
+            $this->symfonyStyle->newLine(3);
+
+            // unset all rectors that are not collector
+            // set new collector rectors - have a custom tag? yes
+        }
+
         // REPORTING PHASE
-        // 3. reporting phase
+        // 4. reporting phaseRunning 2nd time with collectors data
```
Comment (Wirone) [inline src/Console/Command/ProcessCommand.php:106]: This probably should not be here (pasted title from previous phase).
